microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslHandlerBenchmark.java

```diff
@@ -0,0 +1,184 @@
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.ChannelHandler;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.SslProvider;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.microbench.channel.EmbeddedChannelWriteAccumulatingHandlerContext;
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.ReferenceCountUtil;
+import org.openjdk.jmh.annotations.Param;
+
+import java.io.File;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLException;
+
+import static io.netty.handler.codec.ByteToMessageDecoder.COMPOSITE_CUMULATOR;
+import static org.junit.Assert.assertNull;
+
+public class AbstractSslHandlerBenchmark extends AbstractMicrobenchmark {
+    private static final String PROTOCOL_TLS_V1_2 = "TLSv1.2";
+
+    public enum SslEngineProvider {
+        JDK {
+            @Override
+            SslProvider sslProvider() {
+                return SslProvider.JDK;
+            }
+        },
+        OPENSSL {
+            @Override
+            SslProvider sslProvider() {
+                return SslProvider.OPENSSL;
+            }
+        },
+        OPENSSL_REFCNT {
+            @Override
+            SslProvider sslProvider() {
+                return SslProvider.OPENSSL_REFCNT;
+            }
+        };
+        private final SslContext clientContext = newClientContext();
+        private final SslContext serverContext = newServerContext();
+
+        private SslContext newClientContext() {
+            try {
+                return SslContextBuilder.forClient()
+                        .sslProvider(sslProvider())
+                        .trustManager(InsecureTrustManagerFactory.INSTANCE)
+                        .build();
+            } catch (SSLException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        private SslContext newServerContext() {
+            try {
+                File keyFile = new File(getClass().getResource("test_unencrypted.pem").getFile());
+                File crtFile = new File(getClass().getResource("test.crt").getFile());
+
+                return SslContextBuilder.forServer(crtFile, keyFile)
+                        .sslProvider(sslProvider())
+                        .build();
+            } catch (Exception e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        SslHandler newClientHandler(String cipher) {
+            SslHandler handler = clientContext.newHandler(PooledByteBufAllocator.DEFAULT);
+            configureEngine(handler.engine(), cipher);
+            return handler;
+        }
+
+        SslHandler newServerHandler(String cipher) {
+            SslHandler handler = serverContext.newHandler(PooledByteBufAllocator.DEFAULT);
```

Comment:
See my other comments. I think we should create a new instance of the allocator to ensure we not re-use previous filled caches etc.